File: PythonCode/crowdsignals_ch3_outliers.py
##############################################################
#                                                            #
#    Mark Hoogendoorn and Burkhardt Funk (2017)              #
#    Machine Learning for the Quantified Self                #
#    Springer                                                #
#    Chapter 3                                               #
#                                                            #
##############################################################

from util.VisualizeDataset import VisualizeDataset
from Chapter3.OutlierDetection import DistributionBasedOutlierDetection
from Chapter3.OutlierDetection import DistanceBasedOutlierDetection
import copy
import pandas as pd
import numpy as np

# Let is create our visualization class again.
DataViz = VisualizeDataset()

# Read the result from the previous chapter, and make sture the index is of the type datetime.
dataset_path = './intermediate_datafiles/'
dataset = pd.read_csv(dataset_path + 'chapter2_result.csv', index_col=0)
dataset.index = dataset.index.to_datetime()

# Computer the number of milliseconds covered by an instane based on the first two rows
milliseconds_per_instance = (dataset.index[1] - dataset.index[0]).microseconds/1000

# Step 1: Let us see whether we have some outliers we would prefer to remove.

# Determine the columns we want to experiment on.
outlier_columns = ['acc_phone_x', 'light_phone_lux']

# Create the outlier classes.
OutlierDistr = DistributionBasedOutlierDetection()
OutlierDist = DistanceBasedOutlierDetection()

#And investigate the approaches for all relevant attributes.
for col in outlier_columns:
    # And try out all different approaches. Note that we have done some optimization
    # of the parameter values for each of the approaches by visual inspection.
    print '====', col, '===='
    dataset = OutlierDistr.chauvenet(dataset, col)
    DataViz.plot_binary_outliers(dataset, col, col + '_outlier')
    print 'chauvenet outliers ', dataset[col + '_outlier'].sum()
    dataset = OutlierDistr.mixture_model(dataset, col)
    DataViz.plot_dataset(dataset, [col, col + '_mixture'], ['exact','exact'], ['line', 'points'])
    dataset = OutlierDist.simple_distance_based(dataset, [col], 'euclidean', 0.10, 0.99)
    DataViz.plot_binary_outliers(dataset, col, 'simple_dist_outlier')
    print 'simple dist outliers ', dataset['simple_dist_outlier'].sum()
    dataset = OutlierDist.local_outlier_factor(dataset, [col], 'euclidean', 5)
    DataViz.plot_dataset(dataset, [col, 'lof'], ['exact','exact'], ['line', 'points'])

    # Remove all the stuff from the dataset again.
    del dataset[col + '_outlier']
    del dataset[col + '_mixture']
    del dataset['simple_dist_outlier']
    del dataset['lof']

# We take Chauvent's criterion and apply it to all but the label data...

for col in [c for c in dataset.columns if not 'label' in c]:
    print 'Measurement is now: ' , col
    dataset = OutlierDistr.chauvenet(dataset, col)
    dataset.loc[dataset[col + '_outlier'] == True, col] = np.nan
    del dataset[col + '_outlier']

dataset.to_csv(dataset_path + 'chapter3_result_outliers.csv')

File: RCode/crowdsignals_ch3_outliers.R
##########################################################################
#
#   Mark Hoogendoorn & Burkhardt Funk (2017)
#   Machine Learning for the Quantified Self
#   Springer
#   Chapter 3 - Handling Noise and Missing Values in Sensory Data
#   ./crowdsignals_ch3_outliers.R
# 
##########################################################################

# Import libraries and source utils
rm(list=ls())
require(plotly)
require(reshape2)
require(zoo) # used for time series representation
require(KFAS) # Kalman filtering and smoothing
require(Rlof) # package for local outlier factor
source("./chapter2/visiualizeData.R")
source("./chapter3/distributionBasedOutlierDetection.R")
source("./chapter3/distanceBasedOutlierDetection.R")

# Read the result from the previous chapter, and make sture the index is of the type datetime
resultPath = './intermediate_datafiles/'
load(file=paste(resultPath,"chapter2_result.RData",sep=""))

# Compute the number of milliseconds covered by an instane based on the first two rows
timeWindow = as.numeric(difftime(crowdSignalData$time[2],crowdSignalData$time[1],units = "secs"))*1000

# Determine the columns we want to experiment on.
col2Inspect = c('acc_phone_x', 'light_phone_lux')

# Investigate the approaches for all attributes specified in col2Inspect
for (col in col2Inspect) {
  # Note that we have done some optimization of the parameter values for each of the 
  # approaches by visual inspection
  print(paste("====",col,"===="))
  
  # Chauvenet criterion
  print(paste("Chauvenet:",Sys.time()))
  crowdSignalData = chauvenet(crowdSignalData, col)
  plotBinaryOutliers(crowdSignalData,col,paste(col,'_chauvenetOutlier',sep=""))

  # Mixture models
  print(paste("Mixture model:",Sys.time()))
  crowdSignalData = mixtureModel(crowdSignalData,col,numOfComponents = 3, plot = FALSE)
  crowdSignalData[,paste(col,"_mixtureOutlier",sep="")] = crowdSignalData[,paste(col,"_mixtureProb",sep="")]<0.0001
  plotBinaryOutliers(crowdSignalData,col,paste(col,"_mixtureOutlier",sep=""))

  # Distance based approaches require to calculate pair-wise distances, the computation time and
  # the memory requirements increase with N^2, for demonstration purposes we therefore reduce the sample size 
  crowdSignalData = crowdSignalData[1:10000,]
  
  # Simple distance based detection
  print(paste("Simple distance based:",Sys.time()))
  crowdSignalData = simpleDistanceBased(crowdSignalData,col, 0.5, 0.99)
  plotBinaryOutliers(crowdSignalData,col,paste(col,"_simpleDistOutlier",sep=""))
  
  # Local Outlier Factor: we use the Rlof package which implements LOF
  print(paste("Local Outlier Factor:",Sys.time()))
  crowdSignalData = localOutlierFactor(crowdSignalData, col, dFunction = "euclidean", 5)
  crowdSignalData[, paste(col,"_lofOutlier",sep="")] =  crowdSignalData$lof > 5
  plotBinaryOutliers(crowdSignalData, col, paste(col,"_lofOutlier",sep=""))
}

# Reload the result from Chapter 2 and apply Chauvenet criterion to all attributes but the labels
load(file=paste(resultPath,"chapter2_result.RData",sep=""))
col2Inspect = names(crowdSignalData) 
col2Inspect = subset(col2Inspect, (!grepl("label",col2Inspect)&col2Inspect!="time"))
for (col in col2Inspect) {
  
  # Chauvenet criterion
  crowdSignalData = chauvenet(crowdSignalData, col)
  plotBinaryOutliers(crowdSignalData,col,paste(col,'_chauvenetOutlier',sep=""))
  
  # Set outliers to NA
  temp = crowdSignalData[,paste(col,'_chauvenetOutlier',sep="")]
  temp = ifelse(is.na(temp),FALSE,temp)
  crowdSignalData[temp,col] = NA
}

# Delete outlier columns
crowdSignalData = crowdSignalData[,!grepl("_chauvenetOutlier",names(crowdSignalData))]

# Save to file
resultPath = './intermediate_datafiles/'
save(crowdSignalData, file=paste(resultPath,"chapter3_result_outliers.RData",sep=""))